### PR TITLE
Support multiple EID systems for MDM.

### LINKED ISSUE #3173: MDM Multi-Threaded Processing
The HAPI-FHIR side of this is to support multiple MDM EID systems.

### PR DESCRIPTION
**Pull request by tadgh**
Closes #3173 
* deprecate `eidSystem` in favour of `eidSystems`.
* Add implementation, changelog, docs, and tests.
* Add concept of `messageKey` to the base resource message. This can be used by brokers for partitioning purposes.  
* Add concept of `tags` to CanonicalSubscription. These are propagated from the source Subscription Meta tags.

### CODE REVIEW

File: hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/rules/json/MdmRulesJson.java [hunk 1]
```diff
@@ -47,10 +48,16 @@
...
+	 * This field is deprecated, use eidSystems instead.
+	 */
+	@Deprecated
 	@JsonProperty(value = "eidSystem")
 	String myEnterpriseEIDSystem;
 
-
+	@JsonProperty(value = "eidSystems")
+	Map<String, String> myEnterpriseEidSystems = new HashMap<>();
 	@JsonProperty(value = "mdmTypes")
```
Comment (TipzCM) [inline hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/rules/json/MdmRulesJson.java:61]: should this second value be removed?

File: hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/rules/json/MdmRulesJson.java [hunk 2]
```diff
@@ -112,14 +119,57 @@ public void initialize() {
...
+	public void addEnterpriseEIDSystem(String theResourceType, String theEidSystem) {
+		if (myEnterpriseEidSystems == null) {
+			myEnterpriseEidSystems = new HashMap<>();
+		}
+		myEnterpriseEidSystems.put(theResourceType, theEidSystem);
+	}
+
+	public Map<String, String> getEnterpriseEIDSystems() {
+		//First try the new property.
+		if (!myEnterpriseEidSystems.isEmpty()) {
```
Comment (TipzCM) [inline hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/rules/json/MdmRulesJson.java:151]: should we check if null?

File: hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/rules/json/MdmRulesJson.java (hunk 1 above)
Comment (tadgh) [reply hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/rules/json/MdmRulesJson.java:61]: what do you mean second value? the `= new HashMap<>()`?

File: hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/rules/json/MdmRulesJson.java (hunk 2 above)
Comment (tadgh) [reply hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/rules/json/MdmRulesJson.java:151]: Yes, good call.

Comment (tadgh) [reply hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/rules/json/MdmRulesJson.java]: Fixed in 291698381b82ecef89752f5890c5d733d391c4e2

File: hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/rules/json/MdmRulesJson.java (hunk 1 above)
Comment (TipzCM) [reply hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/rules/json/MdmRulesJson.java:63]: nm i misread

**Review by TipzCM — APPROVED** [review]

**Review by katiesmilecdr — APPROVED** [review]

File: hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/rules/config/MdmRuleValidator.java [hunk 3]
```diff
@@ -73,7 +77,40 @@ public void validate(MdmRulesJson theMdmRules) {
...
+		validateSystemsAreUris(theMdmRules);
+		validateEidSystemsMatchMdmTypes(theMdmRules);
+	}
+
+	private void validateEidSystemsMatchMdmTypes(MdmRulesJson theMdmRules) {
+		theMdmRules.getEnterpriseEIDSystems().keySet()
+			.forEach(key -> {
+				//Ensure each key is either * or a valid resource type.
+				if (!key.equalsIgnoreCase("*") && !theMdmRules.getMdmTypes().contains(key)) {
+					throw new ConfigurationException(String.format("There is an eidSystem set for [%s] but that is not one of the mdmTypes. Valid options are [%s].", key, buildValidEidKeysMessage(theMdmRules)));
```
Comment (fil512) [inline hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/rules/config/MdmRuleValidator.java:89]: I gotta remember String.format in exceptions.  This is such a better way to make these errors!

**Review by fil512 — APPROVED** [review]

**Pull request merged**